util/mmio.h

```diff
@@ -69,10 +69,64 @@
 #ifdef __s390x__
 #include <unistd.h>
 #include <sys/syscall.h>
-
-/* s390 requires a privileged instruction to access IO memory, these syscalls
-   perform that instruction using a memory buffer copy semantic.
+#include <sys/auxv.h>
+
+/* s390 requires special instructions to access IO memory. Originally there
+   were only privileged IO instructions that are exposed via special syscalls.
+   Starting with z15 there are also non-privileged memory IO (MIO) instructions
+   we can execute in user-space. Despite the hardware support this requires
+   support in the kernel. If MIO instructions are available is indicated in an
+   ELF hardware capability.
+ */
+extern int s390_mio_supported;
+
+union register_pair {
+	unsigned __int128 pair;
+	struct {
+		uint64_t even;
+		uint64_t odd;
+	};
+};
+
+/* The following pcilgi and pcistgi instructions allow IO memory access from
+   user-space but are only available on z15 and newer.
 */
+static inline uint64_t s390_pcilgi(const void *ioaddr, size_t len)
+{
+	union register_pair ioaddr_len = {.even = (uint64_t)ioaddr, .odd = len};
+	uint64_t val;
+
+	asm volatile (
+		"       .insn   rre,0xb9d60000,%[val],%[ioaddr_len]\n"
```

Comment:
Please put the actualy mmenoic form of the instruction in a comment...

Reply:
Will do